### ISSUE TITLE
Side-trace Russian Roulette

### ISSUE THREAD

**Issue by lukego** [LuaJIT/LuaJIT#218]
This is a test case demonstrating a program that has inconsistent performance due to JIT heuristics. In the future I hope to find ways to make such programs perform more consistently. (More background in #114.)

The [LuaJIT Numerical Computing Performance Guide](http://wiki.luajit.org/Numerical-Computing-Performance-Guide) says:

> Heavily biased branches (>95% in one direction) are fine.

However I think it is more accurate to say that a branch that is 95% biased in one direction is fine _95% of the time_. The caveat is that the branch has to be biased towards taking the root trace instead of a side trace, which is to say that it has to be biased to follow the same branch that was taken in the recording phase. If the code happens to branch in the opposite direction to the bias on the iteration when recording is done then the JIT will optimize for the wrong case and performance will suffer.

The test case is a simple simulation of a million people playing russian roulette. Each person has a 1 in 6 chance to die. The simulation counts the number of people who live and die. The whole simulation is run a thousand times and we record the efficiency (total instructions executed).

```
#!/bin/sh
for i in $(seq 1 1000); do
   perf stat -e instructions luajit - 2>>perf.log <<EOF
      math.randomseed($i)
      local live = 0
      local die  = 0
      for i = 1, 1e6 do
         if math.random(6) == 6 then
            die = die + 1
         else
            live = live + 1
         end
      end
      print(("Survived %.1f%%"):format(live*100/(live+die)))
EOF
done
```

The output of a few iterations looks like this:

```
Survived 83.4%
Survived 83.3%
Survived 83.3%
Survived 83.4%
Survived 83.4%
```

which is consistent with 1 in 6 fatalities (83.333%).

More to the point is that if we look in `perf.log` to see how long each of the 1000 simulation runs took we see that 83.4% of the time it took 67 instructions per iteration and the rest of the time it took 89 instrucitons per iteration. This means that the performance of the simulator as perceived by an end-user is inconsistent (bimodal).

This can be explained as the JIT optimizing either towards or away from the branch's bias depending on the random number that happens to be drawn on the iteration that is recorded.

I am really keen to find practical ways to make code like this always perform consistently from the end-user perspective. This is really important for applications like Snabb that are performance sensitive and where the same trace can run for years at a time as part of a router. It would be very valuable to be able to depend on consistent performance, for example the JIT always discovering the actual bias of a trace, rather than having to model and explain these more exotic outcomes (such as our CI that is now running thousands of tests and presenting the results as [probability density functions](https://hydra.snabb.co/build/589970/download/2/report.html)).

I realize that Hyperblock Scheduling has been identified as the Right Thing for making side-traces more efficient but I am hoping to identify a worse-is-better solution (perhaps inspired by CPU branch prediction heuristics in which simple dynamic mechanisms can be surprisingly effective).

To be continued...

**Comment by corsix** [comment]
> I am really keen to find practical ways to make code like this always perform consistently from the end-user perspective.

To throw one idea into the ring: add a saturating say 8-bit counter to every root trace, which is set to say 200 when the trace is created, and is manipulated upon trace exit:
  * "Good" exits increment the counter by say 5 (up to its saturation limit of say 255).
  * "Bad" exits decrement the counter by say 1 - unless the counter is zero, in which case the bad exit causes the root trace to be thrown away.

For traces which contain a loop, all exits after the ->LOOP label are good. For root traces which do not contain a loop, the final exit is good. All other exits are bad.

**Comment by lukego** [comment]
(Ideas welcome :-))

**Comment by lukego** [comment]
@corsix Cool idea! I suppose that if you enter a trace with count=0 then you immediately flush it and record a new copy. Seems to have some nice properties:

- This retracing logic will not fire at all for the majority of traces, only on ones where the bias is mispredicted or that are exceptionally branchy.
- Repeated tracing should quickly discover the right trace for biased cases. On this example we would get the right answer 83.3% of the time with one attempt, 97.2% with two attempts, 99.5% with three attempts, etc.

Just want to throw out a couple of ideas for extension, mostly to map out more of the terrain:

Suppose that instead of flushing the old trace we would keep a collection of (say) 4 root traces and we would measure their individual efficiency with an "instructions per iteration" metric. Then we could use an [epsilon-greedy](https://en.wikipedia.org/wiki/Multi-armed_bandit) strategy to balance between sticking with the empirically most efficient trace vs. getting new measurements for another trace vs replacing the slowest trace with a new recording. This could potentially allow us to react quickly to familiar workload changes (without JITing new code).

Then if we had multiple root traces and a strategy for selecting between them we could consider making a separate selection for each call site. (Like when CPU branch predictors take the stack into account.) This could open up new optimization possibilities e.g. for code like this:

```lua
loop(500, 'mode1')
loop(1000, 'mode2')
loop(500, 'mode1')
```

On the one hand the `loop` trace will look unbiased (50/50 mode1 vs mode2) when using only one root trace. On the other hand it will look 100% biased if we had separate root traces for mode1 and mode2 and could select the right one for each call site.

I am not sure whether these ideas would carry enough benefit for the complexity. I wonder though, would there be any particular challenges to supporting multiple root traces in parallel?

**Comment by lukego** [comment]
Cool! I found another good test case: the scimark MC (Monte Carlo) benchmark from the [test suite](https://github.com/LuaJIT/LuaJIT-test/cleanup). This benchmark is sensitive to non-deterministic JIT heuristics due to an [inner loop with a biased branch](https://github.com/LuaJIT/LuaJIT-test-cleanup/blob/master/bench/scimark-2010-12-20.lua#L220-L224). Running this benchmark with LuaJIT v2.1 (63465fe71d2605499ce5f3355db3629fb8283cb2) gives me two different scores: 75% of the time I see 484 and 25% of the time I see 221. This seems to be cutting performance to less than half.

I believe the issue is that the `if` branch in the inner loop is biased and the JIT does not always choose the most common case for the root trace. Profiling with traceprof (https://github.com/snabbco/snabb/pull/623) seems to confirm.

Fast case:

```
traceprof report (recorded 4470/4470 samples):
 40% TRACE   7:LOOP          ->loop    scimark.lua:220
 40% TRACE   7               ->loop    scimark.lua:220
 15% TRACE   8      (7/7)    ->7       scimark.lua:220
  3% TRACE   9      (7/3)    ->7       scimark.lua:220
```

Slow case:

```
traceprof report (recorded 6208/6208 samples):
 67% TRACE   7               ->loop    scimark.lua:220
 23% TRACE   7:LOOP          ->loop    scimark.lua:220
  7% TRACE   8      (7/3)    ->7       scimark.lua:223
  1% TRACE   9      (7/6)    ->7       scimark.lua:223
```

This says that on the fast run the CPU hotspot is the looping part of the trace (`7:LOOP`) while in the slow case more time is spent in the non-looping part of the trace (`7`). This makes sense if side-traces are frequently being taken and the JIT has to repeat the setup work when rejoining the root trace (lukego/blog#8).

Caveat: I did have to patch the scimark MC benchmark to provide some entropy to its random number generation. The benchmark in the repo uses a constant random seed that never exercises the bad case. Here is the patch I used:

```diff
diff --git a/bench/scimark-2010-12-20.lua b/bench/scimark-2010-12-20.lua
index 353acb7..15e7cf7 100644
--- a/bench/scimark-2010-12-20.lua
+++ b/bench/scimark-2010-12-20.lua
@@ -8,9 +8,9 @@

 local SCIMARK_VERSION = "2010-12-10"
 local SCIMARK_COPYRIGHT = "Copyright (C) 2006-2010 Mike Pall"
-
+math.randomseed(os.time())
 local MIN_TIME = 2.0
-local RANDOM_SEED = 101009 -- Must be odd.
+local RANDOM_SEED = math.random(100000)*2+1 -- Must be odd.
 local SIZE_SELECT = "small"

 local benchmarks = {
```

**Comment by lukego** [comment]
Just thinking aloud to try and formulate the problem better... please correct my misconceptions!

Background: LuaJIT compiles loops into collections of _N_ traces. Each trace corresponds with one execution path in the source code: whether an `if` takes its `then` or `else` arm, whether an `and` short-circuits or executes its right-hand argument, and so on.

For example here is the code for a loop with three executions paths, each of which would compile into a separate trace:

```lua
for i = 1, n do 
   ...
   if     i % 5 == 0 then
      ...
   elseif i % 10 == 0 then
      ...
   else
      ...
   end
end
```

LuaJIT heuristically chooses one of these traces (executions paths) to be the _root trace_ which is optimized more than the others by hoisting loop-invariant code out of the loop. (Is there a specific name for this optimization?) The other traces are called _side traces_ and are less efficient because they must re-execute all loop-invariant code on each iteration. (How is _side-trace_ defined exactly? Does it also include related traces that are not part of a loop? Can a side-trace also have a loop of its own?)

Overall performance is best when the root trace represents the most frequently taken execution path. That is because this minimizes the average number of times that the loop-invariant code is executed. Conversely, performance suffers if the most frequently taken execution paths are represented by side-traces.

"The problem" is to formulate a dependable heuristic for choosing which execution path should be optimized as the root trace. Currently the heuristic is first-come-best-dressed: the JIT picks one iteration arbitrarily and follows its execution to define the root trace. The following executions are compiled to side-traces. The lifetime of these decisions is indefinite, e.g. the trace mapping may be chosen during the first millisecond of execution and then stick for several years in a long-lived process.

So how should we reformulate this heuristic? (Ideally in such a way that the formulation corresponds to some reasonable implementation strategy?) Here are a few ideas for invariants we may want to establish:

- The root trace is always the code path that has been taken most frequently.
- Weaker alternative: The root trace is never mapped to a code path that is taken less frequently that the average for all code paths.
- Perhaps orthogonally: The trace structure is always determined by behavior within a well-defined window of time e.g. the past <N> milliseconds/seconds/minute or iterations of the loop. (To ensure that long-lived processes react to changes in workload e.g. server applications that may be started during a low-traffic period and would benefit from being re-JITed after workload changes.)

Coming back to @corsix ideas:

- Are we concerned about non-looping traces? Why?
- Can we use counters on trace exits to estimate the relatively frequency of the root trace being taken? (I suppose the direct way to measure this would involve putting a counter into the inner loop of the root trace instead but that sounds expensive and ugly.)
- Are there some bad cases that would cause "thrashing"? If so then should we use a throttling/backoff mechanism to contain this?

How am I doing?

**Comment by lukego** [comment]
Related idea: Could we make side-traces branch back to _themselves_ rather than to the root trace? That is, for each side-trace to contain a complete copy of the loop but optimized for a different code path? This would seem to optimize for the case where a new iteration takes the same path as the previous one. This may be a relatively practical and predictable semantics. (I have a vague idea that @mraleph alluded to this at one point but I may have misunderstood...)

**Comment by lukego** [comment]
Oh! Hey! Cool! Here is an idea for a design that might work well for long-lived processes and also be really easy to implement (no compiler changes):

Use a sampling profiler to identify "suspicious" traces that should be flushed. A trace is suspicious if it is a root trace that spends less than 1/3 of its time in the looping part.

Implementation:
- Add 8-bit counter to each root trace.
- Add SIGPROF hook running at 1kHz. If this fires in the non-looping part of a root trace then increment counter by 1. If it fires in the looping part of a root trace then decrement the counter by 2. If the counter overflows then flag the trace for flushing.
- Have a hook somewhere that detects when a trace is flagged to be flushed and does so. (Hopefully there is a suitable place for such a hook already but otherwise it could be polled manually with a `jit.autoflush()` function.)

That's it. This would be basically automating the post-hoc analysis that I do with traceprof. Looking again at the traceprof examples above, in the case where code was running fast we had an equal amount of time spent in the looping and non-looping parts of the trace:

```
 40% TRACE   7:LOOP          ->loop    scimark.lua:220
 40% TRACE   7               ->loop    scimark.lua:220
```

while in the slow case where the root trace went against the bias this trace spent only 1/4 of its time in the looping part:

```
 67% TRACE   7               ->loop    scimark.lua:220
 23% TRACE   7:LOOP          ->loop    scimark.lua:220
```

So the algorithm would leave the first instance alone but flag the second for flushing.

Could work, no? Just would need a little protection about bad cases like thrashing when the best trace available still qualifies as suspicious, could use exponential backoff based on wall-clock time for example to not flush any one trace too frequently. (also kind of nice that the nature of sampling profiling means the frequency that a trace is flushed will be proportional to how much CPU it consumes i.e. we will mostly ignore traces that are not relevant to overall performance.)

**Comment by lukego** [comment]
I have implemented the LuaJIT side of this scheme over on #224. Now we can use the `jit.util` introspection API to check the ratio of time spent in the loop vs the nonloop code.

I made a little [self-profiling demo program](https://gist.github.com/lukego/458e710275fcf028377dc22e358f8772) that runs the `roulette` benchmark 10 times, forcing a new trace each time, and the prints a table summarizing how each trace executed. Here is an example output:

```
roottrace  nonloop     loop    looping suspicious
2               40     39.0        49% 
6              150     29.0        16% *
10              36     42.0        53% 
14              39     39.0        50% 
18              37     37.0        50% 
22              39     56.0        58% 
26             152     23.0        13% *
30              32     47.0        59% 
34              38     48.0        55% 
38              39     46.0        54% 
```

Here we see that in 8/10 cases we are spending about half of our time in the loop, but in the other 2/10 "suspicious" cases we are spending less than 20% of our time in the loop. The latter are the slow and problematic traces where the bias does not match the root trace.

The next logical step is for the program to periodically audit its own traces and flush any suspicious traces that it finds. This would be a self-healing program that reliably matches biased branches to the root trace. I will write such a function for Snabb and check for an impact on our end-to-end benchmarks.

**Comment by lukego** [comment]
Alas, this approach of flushing pre-loop-heavy traces will not be sufficient in practice. The "gotcha" I have quickly discovered in testing with Snabb is the risk that a loop in the source code does not form a looping trace at all. For example, depending on heuristics during the recording phase, a source code loop could form a looping trace, or could form non-looping traces, or could fall back to interpretation. This mechanism only detects looping traces that are inefficient and is oblivious to cases where a loop was not formed at all.

So, back to the drawing board, but with more visibility thanks to the #224 traceprofile feature.

Couple of immediate thoughts:

Could be potential in a more coarse approach of evaluating the overall process behavior and doing a global `jit.flush()` when it is inadequate.

Could it make sense to explicitly measure loop bias in the pre-recording phase? That is, if the loop becomes hot on the 30th iterations and gets recorded, perhaps we could use information from the first 29 iterations to abort recording if we seem to be tracing away from the bias? This would be to add a bias-estimator similar to the existing hotness-estimator. (Such a mechanism would seem to stay in the spirit of CPU branch predictors.) If this idea is not completely crazy, I wonder how one might implement it...

Just for fun here are a few "screenshots" of whole process traceprofiles. This is a bit like watching `vmstat` on a Unix machine. The first two and second two are each running the same workload. We see major differences in runtime behavior and AFAIK this is purely down to unpredictable JIT heuristics.

```
  Interp   C (LJ)       GC     Exit   Record      Opt      ASM  Nonloop     Loop  Foreign
 22.419%  21.722%  13.550%   1.564%   1.028%   0.011%   0.386%   2.539%  18.327%  18.455%

  Interp   C (LJ)       GC     Exit   Record      Opt      ASM  Nonloop     Loop  Foreign
  2.505%   1.919%   1.583%   4.240%   0.716%   0.000%   0.889%   5.313%  35.567%  47.267%

  Interp   C (LJ)       GC     Exit   Record      Opt      ASM  Nonloop     Loop  Foreign
  0.124%   0.094%  17.859%   0.003%   0.103%   0.000%   0.009%   4.585%  60.022%  17.201%

  Interp   C (LJ)       GC     Exit   Record      Opt      ASM  Nonloop     Loop  Foreign
  0.063%   0.109%   3.150%   0.000%   0.143%   0.000%   0.008%  13.631%  64.582%  18.314%
```

Gonna need some nice tools to dig in more deeply!

Relatedly, this code does involve one very long and performance-critical function that may be written in a programming style that is more-or-less pessimal for a tracing JIT. So another legitimate outcome of all this work could be to succinctly identify the subset or programming styles that can be JITed predictably. (Bonus points for identifying idioms where the JIT will outperform a static compiler due to having more information available at runtime e.g. values of configuration parameters.)

**Comment by lukego** [comment]
Just one more idea to throw out...

Jim Gray wrote a classic paper called [Why do computers stop and what can be done about it?](http://www.hpl.hp.com/techreports/tandem/TR-85.7.pdf). One observation is that an important class of error, total failure at unpredictable times, can be adequately taken care of with redundancy. If you have something important that is prone to failure then this is not a practical problem provided that you are able to smoothly switch to a backup at the appropriate time.

Could we reasonably keep a backup JIT state? That is, if I detect that the current JIT state is suspicious, could I create a new one with `jit.flush()` but somehow reserve the ability to revert to the previous one if the new one is worse? This could make it more practical to search JIT-space for the traces that perform best for the current workload.

**Comment by lukego** [comment]
Just a little more musing...

I came upon an analogy that I find encouraging: tracing JIT heuristics are a lot like TCP transmission heuristics. For example, your JIT may handle biased branches well 95% of the time have edge cases like anomalous branches during recording, and your TCP may tolerate minor packet loss very well except for edge cases like losing a packet very early in a short-lived connection.

TCP has evolved very well over the years. The best modern implementations handle all manner of different conditions and very seldom make mistakes. However, it has been a long process to find robust heuristics to respond well to network congestion, bit errors, packet reordering, queueing delay, intermittent connectivity, and other circumstances that have evolved as networks become faster / go cellular / go satellite / etc.

So on reflection I am quite enthusiastic about a long journey of incrementally identifying bad cases, tweaking heuristics to handle them better, and gradually replacing complicated special-case heuristics with simple general ones as they become apparent. I am not really tempted to go back to static complication (or circuit-switched telephony).

Onward :-) :beers:

**Comment by DemiMarie** [comment]
I wish that someone could implement hyperblock scheduling.  That would solve all of the problems by virtually eliminating the penalty for branches in loops (beyond that from CPU branch prediction).

**Comment by mejedi** [comment]
>Alas, this approach of flushing pre-loop-heavy traces will not be sufficient in practice. The "gotcha" I have quickly discovered in testing with Snabb is the risk that a loop in the source code does not form a looping trace at all. For example, depending on heuristics during the recording phase, a source code loop could form a looping trace, or could form non-looping traces, or could fall back to interpretation. 

The difference between looping traces and side traces reentering the same loop is due to loop invariant code getting hoisted out of a loop (in looping traces) which leads to noticeable performance improvements. So if a loop in source code produces a non-looping trace it's largely irrelevant since the optimisation doesn't apply anyway.

Ex: an outter loop, assuming the inner loop was traced, produces a side trace emerging from the inner loop's trace guard validating the loop termination condition and eventually transfers the control back to the inner loop's mcode.

Now jumping from trace to trace has some inherent overhead. Almost every bytecode instruction updates a stack slot it refers to. Ex: `sum=sum+i` translates to `ADDVV 0 0 4` assuming `sum` was assigned slot 0 and `i` got 4. Maintaining these semantics in compiled code would be a waste, hence in compiled code updating stack slots is often deferred until trace exit. There are so called *snapshots* telling how to re-create correct Lua stack on particular trace exit.

When a side trace branches out it *inherits* the respective snapshot. In particular it may benefit from having certain values readily available in registers and often may omit type checks. But when a side trace eventually links to another trace it's a different story. At this point the Lua stack must be re-created, and then the trace we linked to reads values from the stack and typechecks everything again. This is because currently there's no way to tell whether a (root) trace is entered from the interpretor or if another trace invoked it.

I feel that designing an efficient linking mechanism that could bypass redundant typechecks and, ideally, even skip loop invariant code hoisted out of a loop will solve the problem at hand and is much more tangible than the almighty hyperblock scheduling :)

One final note on flushing the bad traces. The problem is that it's currently not possible to purge individual traces. I.e. we can sever the link between the bytecode and the trace so that it's not going to be used, however the trace still counts towards the trace limit, which is 1000 by default (could be raised as high as 65K). Once the limit is hit, *all* traces currently in existence are purged.

**Comment by lukego** [comment]
@mejedi This is very much along the lines I am thinking too :-)

Taking a step further back, I actually wonder if there are a few related areas that could be optimized in a complementary way:

1. Cost of branching between traces. For example, making it cheaper for side-traces to rejoin their parent for a loop.
2. Frequency of branching between traces. For example, allowing more side traces to form their own local loops instead of rejoining the parents.
3. Cost of loop-invariant code. The main difference between a looping trace and a non-looping side-trace is that the latter has to rerun all of the loop-invariant instructions each time it rejoins the parent. So it loses the benefit of "loop optimization" that otherwise hoists these out of the loop body. So: Could we make the loop-invariant code more efficient for the cases where it needs to be repeated? This would reduce the performance delta between looping and non-looping traces. Could be that we could hoist from loop-invariant code to run at _recording time_ (#248) so that things like looking up modules in the _G dictionary, etc, are free for both looping and non-looping traces.

On this note I would like to take a closer look at typical loop-invariant code and think about how to optimize it rather than only how to bypass it.

Relatedly: I am having some success lately with avoiding side-trace penalties by forcing the JIT to create multiple root traces. This can be very effective when you know what trace behavior you want and are able to communicate it to the JIT. Example from yesterday is https://github.com/rtsisyk/luafun/pull/33.

**Comment by lukego** [comment]
I'm rereading the IR code for the [parent and side trace](https://gist.github.com/lukego/609b4e61b28c6666332e) of lukego/blog#8 more closely now. Seems like the relationship between side traces and parent traces is more nebulous than I had credited. In particular the side-trace seems to be repeating some loop-invariant work that the parent trace had already taken care of. So for example it seems like looking up the `counter.add` function in the `counter` hashtable is done _twice_ on some iterations: first in the parent trace and then again in the side trace.

This could potentially make optimizations like @fsfod's "read-only tables" twice as valuable: if you can take care of some code at recording time then you skip running it in both the parent trace and the side traces.

I need to spend more brain cycles noodling on the idea of moving more computation to recording time. This seems like a potential big win. The other big win I see is like in the luafun example where you avoid excessive sharing i.e. if you have a function that is used in <n> different ways (e.g. different function parameters, different loop biases, etc) then you arrange to have <n> separate traces and to select the right one.

Fun topic!

**Comment by lukego** [comment]
(Closed to avoid clutter.)
